### PR TITLE
ENH: Add vtkMRMLTextNode and "Texts" module

### PR DESCRIPTION
**Pull request by Sunderlandkyl**
vtkMRMLTextNode and vtkMRMLTextStorageNode migrated to Slicer core from SlicerOpenIGTLink.

The Texts module adds a new designer plugin widget: "qMRMLTextWidget".
qMRMLText provides an text box with properties for readOnly, and continuousUpdate.

image

### CODE REVIEW

**Comment by Sunderlandkyl** [comment]
- Added vtkMRMLTextNode and vtkMRMLTextStorageNode tests.
- Added warning messages when saving/cancelling an edit in qMRMLTextWidget could overwrite/discard data:
image 
image

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (jcfr) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: What do you think of using `autoUpdate` ?  This would be consistent with the  `updateWidget` and `updateFinished` method names

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (jcfr) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: onEditClicked -> onEditButtonClicked ? Same comment for Cancel and Save

That said, I wonder if it makes sense to have these signals. Having signal like `updateRequested` and `updateCanceled` would be more generic  and would also nicely complement the existing `updateFinished` signal

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (jcfr) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]:
```
\sa setMRMLNode, mrmlNode()
```

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (jcfr) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Comment should be updated

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (jcfr) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: propery -> property

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (jcfr) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Instead of having a signal in the widget, should we instead have a a property in the text node ?

File: Libs/MRML/Core/vtkMRMLTextNode.h
Comment (jcfr) [inline Libs/MRML/Core/vtkMRMLTextNode.h]: It would be nice to also support UTF8 and add test for it.

**Review by jcfr — CHANGES_REQUESTED** [review]
This is great ! Thanks for working on this. 

Having first class support for text base document like json, xml, txt is very nice.

I wonder if this should be used as a base for other node stored as text or json (e.g markup)

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Sounds good! autoUpdate() is much better than continuousUpdate().

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: I think that they would serve two different purposes, although having a readOnly property in the text node does seem like a useful feature.

- Read only in the qMRMLTextWidget allows text node content display without allowing editing (Whether or not the node is read only).

File: Libs/MRML/Core/vtkMRMLTextNode.cxx
Comment (Sunderlandkyl) [reply Libs/MRML/Core/vtkMRMLTextNode.cxx]: Changed const* char to std::string.

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Done!

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Changed methods to XButtonClicked(). Added new signals (updateRequested, editingChanged(bool)).

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Done!

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Left ReadOnly out of the node for now.

Re a discussion with @lassoan, If it was implemented, it would probably be like the Locked parameter in markup and table nodes.

File: Libs/MRML/Core/vtkMRMLTextNode.h
Comment (Sunderlandkyl) [reply Libs/MRML/Core/vtkMRMLTextNode.h]: Added a test for UTF-8, and decided to just use the VTK_ENCODING_X enum to represent encoding.

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Done!

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Done!

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Done!

File: Libs/MRML/Core/vtkMRMLTextNode.cxx
Comment (lassoan) [inline Libs/MRML/Core/vtkMRMLTextNode.cxx]: It would be better to use the variable name ForceCreateStorageNode with enum value: CreateStorageNodeAuto, CreateStorageNodeAlways, CreateStorageNodeNever (to make the behavior more predictable).

File: Libs/MRML/Core/vtkMRMLTextNode.cxx
Comment (lassoan) [inline Libs/MRML/Core/vtkMRMLTextNode.cxx]: Performance could be slightly better with const std::string &text (which might matter if the string is long and updates are frequent). If you don't see any drawback then use const ref instead.

File: Libs/MRML/Core/vtkMRMLTextNode.cxx
Comment (lassoan) [inline Libs/MRML/Core/vtkMRMLTextNode.cxx]: MAX_STRING_LENGTH_FOR_SAVE_WITHOUT_STORAGE_NODE

File: Libs/MRML/Core/vtkMRMLTextNode.h
Comment (lassoan) [inline Libs/MRML/Core/vtkMRMLTextNode.h]: It is interesting that this compiled (const is in the header but not in the implementation). Anyway, it probably should be const std::string&.

File: Libs/MRML/Core/vtkMRMLTextNode.h
Comment (lassoan) [inline Libs/MRML/Core/vtkMRMLTextNode.h]: Probably we should invoke TextModifiedEvent even if encoding was changed.

File: Libs/MRML/Core/vtkMRMLTextNode.h
Comment (lassoan) [inline Libs/MRML/Core/vtkMRMLTextNode.h]: Document when this event is invoked. Also document if Modified event is not invoked for certain Set methods.

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx
Comment (lassoan) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx]: Direction of update is not clear. Use updateWidgetFromMRML (updateWidgetFromMRMLRequested).

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx
Comment (lassoan) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx]: What happens if you set read-only while the widget is being edited? Do you save the results? (probably should)

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx
Comment (lassoan) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx]: Proper QMessageBox deleting is tricky. You may need to make it auto-close if you don't destroy explicitly. Search for other similar messageboxes and copy their implementation.

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx
Comment (lassoan) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx]: Mention that the text has been modified _since editing was started_.

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (lassoan) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Add a note on how auto-update works while editing is in progress.

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (lassoan) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: Add some more info about this. How the appearance and behavior is different if read-only or not.

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h
Comment (lassoan) [inline Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h]: rename to update...From...Requested

File: Modules/Loadable/Texts/qSlicerTextsModule.cxx
Comment (lassoan) [inline Modules/Loadable/Texts/qSlicerTextsModule.cxx]: Put it in the same category as Tables (Informatics).

**Review by lassoan — COMMENTED** [review]
Very nice, thank you. I just added a couple of small comments that should be mostly trivial to address.

File: Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx
Comment (Sunderlandkyl) [reply Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx]: This is how QMessageBox and ctkMessageBox are handled in qMRMLSceneViewMenu and qMRMLAnnotationTreeView.

Is there another style of implementation that would be better?

**Comment by Sunderlandkyl** [comment]
I made most of the changes (With the exception of the QMessageBox change), and added some more lines in the test for the new storage node options (CreateStorageNodeAlways, CreateStorageNodeAuto, CreateStorageNodeNever).

**Comment by Sunderlandkyl** [comment]
Message boxes removed in favor of a warning label under the save button:
image

**Comment by lassoan** [comment]
Great, thank you. Maybe we could make it more subtle? For example, add a warning icon to the "Save" button and add the "The original text has been modified..." text to the button's tooltip. This way, we would not need to change the size of the widget and the warning icon may be more visible then the text below.

**Comment by Sunderlandkyl** [comment]
Changed to icon+tooltip on save button:

image

**Comment by lassoan** [comment]
Thank you, this looks good. I'll merge this next week after the CARS tutorial is finished, just to make sure we don't introduce any regressions in the last minute.

**Comment by lassoan** [comment]
I've reviewed and tested this and pushed some fixes/improvements. There are just a few things that would be nice to address before merging:
- [ ] Clear widget contents when scene is closed.
- [ ] Add word wrap option (Boolean flag in the text edit widget and a checkbox in the module widget's advanced section)
- [ ] Double-check that Qt designer plugin works (for me the text edit widget did not show up in QtDesigner, but maybe I just did not rebuild enough)

**Comment by Sunderlandkyl** [comment]
Text nodes in SH currently show node contents in tooltip:
image

"Edit text" in SH context menu:
image

**Comment by lassoan** [comment]
Looks good!
What does "Edit text" do? If the same as "Edit properties..." (open Texts module with the text node selected) then it is probably redundant.

**Comment by Sunderlandkyl** [comment]
Yeah it does. Didn't even notice that "Edit properties..." was there.

**Comment by Sunderlandkyl** [comment]
Removed "Edit text" and added encoding info to the SH tooltip.

**Comment by lassoan** [comment]
@Sunderlandkyl 

> added encoding info to the SH tooltip.

I'm not sure if this is needed. Encoding is only needed to interpret the byte stream and convert it to displayable string. If we already display the string in the tooltip it does not really matter what encoding was used (especially not for users). If the text is empty or very short then displaying encoding information might be distracting.

Let me know if you could address the other things (scene close, etc.) and it is ready for testing. Thank you!

@jcfr 

> I wonder if this should be used as a base for other node stored as text or json (e.g markup)

This is a good idea, but we need to think about it a bit more. This pull request already contains many changes, so I would merge this and get back to this idea later.

**Comment by Sunderlandkyl** [comment]
Sorry, the other issues have been addressed.

What do you think the ideal tooltip should be?

1.
```
Text (Encoding: ASCII)
// Node contents
```
2.
```
Text
// Node contents
```
3.
```
// Node contents
```

1 and 2 are more consistent with other node types, while is much simpler.

**Comment by lassoan** [comment]
I like both 1 and 3 (1 is more consistent with other nodes; 3 is more to the point). Let's just keep the current implementation then. We can tune later based on real-world experience. I'll merge this now.

**Comment by lassoan** [comment]
Merged in r28328. Thanks a lot for the great work!

You can update SlicerOpenIGTLink now to not use its own text node in Slicer-4.11.

**Comment by Sunderlandkyl** [comment]
Great, thanks Andras!

**Comment by jcfr** [comment]
Currently working on GitHub transtion, data migration, ... I added a note to the authorship page to add you back as author with Andras as co-author and committer. See https://www.slicer.org/wiki/Documentation/Labs/TransitionToGit#Authorship

**Pull request closed without merging**
